File: apriltag/src/test/native/cpp/proto/AprilTagFieldLayoutProtoTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <vector>

#include <catch2/catch_test_macros.hpp>

#include "wpi/apriltag/AprilTagFieldLayout.hpp"
#include "wpi/util/SmallVector.hpp"

using namespace wpi::apriltag;

namespace {

const AprilTagFieldLayout kExpectedData = AprilTagFieldLayout{
    std::vector{AprilTag{1, wpi::math::Pose3d{}},
                AprilTag{3, wpi::math::Pose3d{0_m, 1_m, 0_m,
                                              wpi::math::Rotation3d{
                                                  0_deg, 0_deg, 0_deg}}}},
    54_ft, 27_ft};
}  // namespace

TEST_CASE("AprilTagFieldLayoutProtoTest Roundtrip", "[apriltag]") {
  wpi::util::ProtobufMessage<decltype(kExpectedData)> message;
  wpi::util::SmallVector<uint8_t, 256> buf;

  REQUIRE(message.Pack(buf, kExpectedData));
  auto unpacked_data = message.Unpack(buf);
  REQUIRE(unpacked_data.has_value());

  REQUIRE(unpacked_data->GetFieldLength() == kExpectedData.GetFieldLength());
  REQUIRE(unpacked_data->GetFieldWidth() == kExpectedData.GetFieldWidth());

  REQUIRE(unpacked_data->GetTagMap() == kExpectedData.GetTagMap());
}

File: apriltag/src/test/java/org/wpilib/vision/apriltag/proto/AprilTagFieldLayoutProtoTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.vision.apriltag.proto;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;
import org.junit.jupiter.api.Test;
import org.wpilib.math.geometry.Pose3d;
import org.wpilib.math.geometry.Rotation3d;
import org.wpilib.math.util.Units;
import org.wpilib.vision.apriltag.AprilTag;
import org.wpilib.vision.apriltag.AprilTagFieldLayout;

class AprilTagFieldLayoutProtoTest {
  private static final AprilTagFieldLayout DATA =
      new AprilTagFieldLayout(
          List.of(
              new AprilTag(1, Pose3d.kZero),
              new AprilTag(3, new Pose3d(0, 1, 0, Rotation3d.kZero))),
          Units.feetToMeters(54.0),
          Units.feetToMeters(27.0));

  @Test
  void testRoundtrip() {
    ProtobufAprilTagFieldLayout proto = AprilTagFieldLayout.proto.createMessage();
    AprilTagFieldLayout.proto.pack(proto, DATA);

    AprilTagFieldLayout data = AprilTagFieldLayout.proto.unpack(proto);
    assertEquals(DATA.getTagMap(), data.getTagMap());
    assertEquals(DATA.getFieldLength(), data.getFieldLength());
    assertEquals(DATA.getFieldWidth(), data.getFieldWidth());
  }
}
